### PR TITLE
[tests] Update Typescript, Jest, and ts-jest

### PR DESCRIPTION
**Pull request by cb1kenobi**
Updates TypeScript, Jest, and ts-jest to newer versions which have improved support for ESM.

This PR contains a subset of changes from https://github.com/vercel/vercel/pull/9692.

### CODE REVIEW

**Review by styfle — DISMISSED** [review]

File: packages/edge/package.json [hunk 1]
```diff
@@ -33,11 +33,14 @@
   "jest": {
```
Comment (styfle) [inline packages/edge/package.json]: In a future PR, I would like to move these all out of package.json and use the `jest.config.js` like we are using in the other packages in this monorepo

Comment (cb1kenobi) [reply packages/edge/package.json]: Done!

**Review by styfle — DISMISSED** [review]
Thanks!

**Review by TooTallNate — APPROVED** [review]

File: packages/cli/test/unit/commands/pull.test.ts [hunk 2]
```diff
@@ -98,10 +98,10 @@ describe('pull', () => {
 
       const config = await fs.readJSON(path.join(cwd, '.vercel/project.json'));
       expect(config).toMatchInlineSnapshot(`
-        Object {
```
Comment (EndangeredMassa) [inline packages/cli/test/unit/commands/pull.test.ts:101]: **question (non-blocking):** ‏was this a different between versions of jest?

File: packages/static-build/test/gatsby.test.ts [hunk 3]
```diff
@@ -281,7 +281,7 @@ describe('gatsby utilities', () => {
     `);
     expect(gatsbyConfig).toMatchInlineSnapshot(`
       "// This file was generated by @vercel/static-build
-      import userConfig from \\"./gatsby-config.mjs.__vercel_builder_backup__.mjs\\";
```
Comment (EndangeredMassa) [inline packages/static-build/test/gatsby.test.ts:284]: **question (non-blocking):** ‏was this also a difference between versions of jest?

**Review by EndangeredMassa — APPROVED** [review]

File: packages/static-build/test/gatsby.test.ts (hunk 3 above)
Comment (cb1kenobi) [reply packages/static-build/test/gatsby.test.ts:284]: I'm not 100% certain, but I'm guessing it was a breaking change in v29.0.0: https://github.com/facebook/jest/blob/main/CHANGELOG.md#2900

```
[jest-config] [BREAKING] Make snapshotFormat default to escapeString: false and printBasicPrototype: false (https://github.com/facebook/jest/pull/13036)
```

File: packages/cli/test/unit/commands/pull.test.ts (hunk 2 above)
Comment (styfle) [reply packages/cli/test/unit/commands/pull.test.ts:101]: I might actually start using snapshots now that the syntax is fixed 😌

**Review by styfle — APPROVED** [review]

File: packages/cli/test/unit/commands/pull.test.ts (hunk 2 above)
Comment (cb1kenobi) [reply packages/cli/test/unit/commands/pull.test.ts:101]: No kidding!

**Pull request merged**